pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+    use frame_support::{
+        ensure, fail, log,
+        pallet_prelude::{
+            DispatchClass, DispatchResult, IsType, OptionQuery, StorageMap, StorageValue,
+            ValueQuery,
+        },
+        traits::{Get, UnixTime},
+        Identity,
+    };
+    use frame_system::{ensure_root, ensure_signed, pallet_prelude::OriginFor};
+    use tendermint::Time;
+    use tendermint_light_client_verifier::{
+        options::Options, types::LightBlock, ProdVerifier, Verifier,
+    };
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// ubiquitous event type
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Maximal number of finalized headers to keep in the storage, last-in first-out
+        #[pallet::constant]
+        type HeadersToKeep: Get<u32>;
+
+        /// time provider type, used to gauge whether blocks are within the trusting period
+        type TimeProvider: UnixTime;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Pallet is halted
+        LightClientHalted,
+        /// Pallet operations are resumed        
+        LightClientResumed,
+        /// Light client is initialized
+        LightClientInitialized,
+        /// New block has been verified and imported into storage \[relayer_address, imported_block_hash\]
+        ImportedLightBlock(T::AccountId, BridgedBlockHash),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Unable to deserialize extrinsic
+        DeserializeError,
+        /// Light client has not been initialized        
+        NotInitialized,
+        /// Light client has already been initialized
+        AlreadyInitialized,
+        /// Light client is currently halted
+        Halted,
+        /// The minimum voting power threshold is not reached, the block cannot be trusted yet
+        NotEnoughTrust,
+        /// Verification failed, the block is invalid.        
+        InvalidBlock,
+    }
+
+    // NOTE for storage:
+    // - header storage should be a ring buffer (i.e. we keep n last headers, ordered by the insertion time)
+    // - we keep a pointer to the last finalized header (to avoid deserializing the whole buffer)
+    // - insertion moves the pointer and updates the buffer
+    // or?
+    // https://substrate.recipes/ringbuffer.html
+
+    /// Hash of the last imported header from the bridged chain
+    #[pallet::storage]
+    #[pallet::getter(fn get_last_imported_hash)]
+    pub type LastImportedHash<T: Config> = StorageValue<_, BridgedBlockHash, ValueQuery>;
+
+    /// A ring buffer of imported hashes "ordered" by their insertion time
+    #[pallet::storage]
+    pub type ImportedHashes<T: Config> = StorageMap<_, Identity, u32, BridgedBlockHash>;
```

Comment:
Specify a getter and use it instead of `ImportedHashes::<T>::get`.

Reply:
done